src/basic/terminal-util.c

```diff
@@ -1842,3 +1842,224 @@ int get_default_background_color(double *ret_red, double *ret_green, double *ret
+                                        context->state = CURSOR_TEXT;
+                                else
+                                        context->row = context->row * 10 + d;
+                        }
+                        break;
+
+                case CURSOR_COLUMN:
+                        if (c == 'R') {
+                                if (context->column > 0) {
+                                        if (ret_processed)
+                                                *ret_processed = i + 1;
+
+                                        return 1; /* success! */
+                                }
+
+                                context->state = CURSOR_TEXT;
+                        } else {
+                                int d = undecchar(c);
+                                if (d < 0 || context->column > (UINT_MAX-d/10))
+                                        context->state = CURSOR_TEXT;
+                                else
+                                        context->column = context->column * 10 + d;
+                        }
+
+                        break;
+                }
+
+                /* Reset any positions we might have picked up */
+                if (IN_SET(context->state, CURSOR_TEXT, CURSOR_ESCAPE))
+                        context->row = context->column = 0;
+        }
+
+        if (ret_processed)
+                *ret_processed = size;
+
+        return 0; /* all good, but not enough data yet */
+}
+
+int terminal_get_size_by_dsr(
+                int input_fd,
+                int output_fd,
+                unsigned *ret_rows,
+                unsigned *ret_columns) {
+
+        assert(input_fd >= 0);
+        assert(output_fd >= 0);
+
+        int r;
+
+        /* Tries to determine the terminal dimension by means of ANSI sequences rather than TIOCGWINSZ
+         * ioctl(). Why bother with this? The ioctl() information is often incorrect on serial terminals
+         * (since there's no handshake or protocol to determine the right dimensions in RS232), but since the
+         * ANSI sequences are interpreted by the final terminal instead of an intermediary tty driver they
+         * should be more accurate.
+         *
+         * Unfortunately there's no direct ANSI sequence to query terminal dimensions. But we can hack around
+         * it: we position the cursor briefly at an absolute location very far down and very far to the
+         * right, and then read back where we actually ended up. Because cursor locations are capped at the
+         * terminal width/height we should then see the right values. In order to not risk integer overflows
+         * in terminal applications we'll use INT16_MAX-1 as location to jump to — hopefully a value that is
+         * large enough for any real-life terminals, but small enough to not overflow anything or be
+         * recognized as a "niche" value. (Note that the dimension fields in "struct winsize" are 16bit only,
+         * too). */
```

Comment:
Why exactly do we have to modify the size in the first place? Aren't we going to get the right size just by querying the size via ANSI instead of the ioctl?

Reply:
Tools such as "less" or "mc" or "top" all check for screen dimensions via TIOCGWINSZ+SIGWINCH, and not via ANSI sequences. That's the key of the problem here. Because that's the way it is, we'll now propagate the info from the ANSI sequences (which is information about the ultimate final terminal) into the data TIOCGWINSZ will return for this tty (which is information from the local tty driver). 

Thus, from that point on "less", "mc", "top" will know the correct dimensions.

Reply:
@poettering I mean, why do we have to change the size via ANSI sequences in the first place? As long as you query the size via ANSI and set that as the tty size in the VM using the ioctl, wouldn't that be sufficient?

Reply:
change the size via ANSI sequence? we do not do that (we can't do that even if we wanted, there's no sequence for that)

hence, we just query via ANSI and set via ioctl, nothing more.

Reply:
Ah it's cursor position that you set not size, nvm I'm just stupid